cosmo_tester/test_suites/summary/__init__.py

```diff
@@ -0,0 +1,599 @@
+                    '{endpoint} did not provide a helpful error message for '
+                    'an invalid field. Error message should have stated '
+                    'which field was invalid, and given a list of fields '
+                    'which were valid. Expected valid fields were: {valid}. '
+                    'Returned message was: {message}'.format(
+                        endpoint=summary_endpoint.summary_type,
+                        valid=', '.join(
+                            SUPPORTED_FIELDS[summary_endpoint.summary_type]
+                        ),
+                        message=str(err),
+                    )
+                )
+
+    if errors:
+        raise AssertionError('\n'.join(errors))
+
+
+def test_bad_subfield_selected(prepared_manager):
+    """
+        Confirm a helpful error message is returned if a bad field is selected
+    """
+    errors = []
+
+    for summary_endpoint in (
+        prepared_manager.client.summary.blueprints,
+        prepared_manager.client.summary.deployments,
+        prepared_manager.client.summary.executions,
+        prepared_manager.client.summary.nodes,
+        prepared_manager.client.summary.node_instances,
+    ):
+        try:
+            summary_endpoint.get(_target_field='tenant_name',
+                                 _sub_field='notafield')
+            errors.append(
+                'Field notafield should not have been accepted by client for '
+                '{endpoint}.'.format(endpoint=summary_endpoint.summary_type)
+            )
+        except CloudifyClientError as err:
+            components = ['notafield', 'not', 'summary', 'Valid',
+                          'tenant_name']
+            if not all(component in str(err) for component in components):
+                errors.append(
+                    '{endpoint} did not provide a helpful error message for '
+                    'an invalid field. Error message should have stated '
+                    'which field was invalid, and given a list of fields '
+                    'which were valid.'.format(
+                        endpoint=summary_endpoint.summary_type,
+                    )
+                )
+
+    if errors:
+        raise AssertionError('\n'.join(errors))
+
+
+def test_basic_summary_blueprints(prepared_manager):
+    results = sorted(prepared_manager.client.summary.blueprints.get(
+        _target_field='tenant_name',
+        _all_tenants=True,
+    ).items)
+
+    expected = sorted([
+        {u'blueprints': 3, u'tenant_name': u'test1'},
+        {u'blueprints': 3, u'tenant_name': u'test2'},
+        {u'blueprints': 3, u'tenant_name': u'default_tenant'},
+    ])
+
+    assert results == expected, (
+        '{0} != {1}'.format(results, expected)
+    )
+
+
+def test_basic_summary_deployments(prepared_manager):
+    results = sorted(prepared_manager.client.summary.deployments.get(
+        _target_field='blueprint_id',
+    ).items)
+
+    expected = sorted([
```

Comment:
this feels wrong with no key func, how do dicts even sort?
(the answer is, they don't, and in py3 this sanely throws an exception instead)

Reply:
Sorry, misunderstood. Either way, with the sorted this manages to be stable but without it doesn't. I could probably put in a key function if it's worthwhile?